File: app/src/main/java/com/zionhuang/music/models/SongParcel.java
package com.zionhuang.music.models;

import android.os.Parcel;
import android.os.Parcelable;

import com.google.api.services.youtube.model.SearchResult;
import com.google.api.services.youtube.model.Video;
import com.zionhuang.music.db.SongEntity;

public class SongParcel implements Parcelable {
    private String id;
    private String title;
    private String artist;

    public SongParcel(String id, String title, String artist) {
        this.id = id;
        this.title = title;
        this.artist = artist;
    }

    public static SongParcel fromSongEntity(SongEntity song) {
        return new SongParcel(song.id, song.title, song.artist);
    }

    public static SongParcel fromVideo(Video video) {
        return new SongParcel(video.getId(), video.getSnippet().getTitle(), video.getSnippet().getChannelTitle());
    }

    public static SongParcel fromSearchResult(SearchResult item) {
        if (!"youtube#video".equals(item.getId().getKind())) {
            throw new IllegalArgumentException("Can't convert a " + item.getId().getKind() + " item to SongParcel.");
        }
        return new SongParcel(item.getId().getVideoId(), item.getSnippet().getTitle(), item.getSnippet().getChannelTitle());
    }

    public String getId() {
        return id;
    }

    public String getTitle() {
        return title;
    }

    public String getArtist() {
        return artist;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(id);
        dest.writeString(title);
        dest.writeString(artist);
    }

    public static final Parcelable.Creator<SongParcel> CREATOR = new Creator<SongParcel>() {
        @Override
        public SongParcel createFromParcel(Parcel source) {
            String id = source.readString();
            String name = source.readString();
            String artist = source.readString();
            return new SongParcel(id, name, artist);
        }

        @Override
        public SongParcel[] newArray(int size) {
            return new SongParcel[size];
        }
    };
}

File: app/src/main/java/com/zionhuang/music/models/SongParcel.kt
package com.zionhuang.music.models

import android.os.Parcelable
import com.google.api.services.youtube.model.SearchResult
import com.google.api.services.youtube.model.Video
import com.zionhuang.music.db.entities.Song
import com.zionhuang.music.extensions.maxResUrl
import com.zionhuang.music.youtube.extractors.YouTubeStreamExtractor
import com.zionhuang.music.youtube.models.YouTubeStream
import kotlinx.parcelize.Parcelize
import org.schabi.newpipe.extractor.stream.StreamInfoItem

@Parcelize
data class SongParcel(
        val id: String,
        val title: String?,
        val artist: String?,
        val channelId: String,
        val channelName: String,
        val artworkUrl: String?,
) : Parcelable {
    companion object {
        fun fromSong(song: Song): SongParcel =
                SongParcel(song.songId, song.title, song.artistName, song.channelId, song.channelName, null)

        fun fromStream(stream: YouTubeStream.Success): SongParcel =
                SongParcel(stream.id, stream.title, null, stream.channelId, stream.channelTitle, stream.thumbnailUrl)

        fun fromVideo(video: Video): SongParcel =
                SongParcel(video.id, video.snippet.title, null, video.snippet.channelId, video.snippet.channelTitle, video.snippet.thumbnails.maxResUrl)

        fun fromSearchResult(item: SearchResult): SongParcel {
            require("youtube#video" == item.id.kind) { "Can't convert a " + item.id.kind + " item to SongParcel." }
            return SongParcel(item.id.videoId, item.snippet.title, null, item.snippet.channelId, item.snippet.channelTitle, item.snippet.thumbnails.maxResUrl)
        }

        fun fromStreamInfoItem(item: StreamInfoItem): SongParcel =
                SongParcel(YouTubeStreamExtractor.extractId(item.url)!!, item.name, "", "", "", item.thumbnailUrl)
    }
}